### DISCUSSION TITLE
Improving Angular 13 Build Times with Dagger

### DISCUSSION THREAD

**Discussion by sheldonhull** [dagger/dagger#4602] [Help] [answered]
This is more a composite of Dagger + Angular knowledge.

- I've got cached volume for both `.angular/cache` and `node_modules`.
- `ng build` doesn't seem too smart on Angular 13 to pickup on no changes.
- I could switch to `nx` which does better, but trying to stick with default toolchain if I can.

The build still takes 3-5 minutes after the first run, not seeming to recognize that the rebuild doesn't really need to happen. Any quick things I can check to improve so the local repeat builds are a bit faster? For this scenario, let's just assume no changes to the Angular project has occurred.

This might be more an Angular question than Dagger so we can close this if my cached volumes are the extent of the optimization I should probably make with Dagger.

**Comment by marcosnils** [comment]
summoning our FE guru's @slumbering and @crjm

**Comment by crjm** [comment, marked as answer]
First of all, sorry for the late reply, was gathering a bit more context before adding my comment. I'm going to answer this in two sections.

I actually never used Angular before and I'm not aware of the granularity the `buildkit` caching actually has/the internal workings of it. However, I have been tinkering with different ways to cache artifacts for a few tools from the JS ecosystem, so the conclusions I have for now should apply to all frameworks/tools from the ecosystem.

## Triggering builds

It's hard to avoid triggering package manager executions, so the best outcomes are when provisioning is done right for each tool, e.g. setting the cache location for each one of them. I think overall this is the best way though, because the tool itself should be smarter about what should be re-run or not. We shouldn't force a cache hit with Dagger if the tool itself can decide whether to run or not. 
For example, when doing a `yarn install`, the command runs, but it returns an `Already up-to-date.` message if the cache folder is correctly provisioned to the tool. This is the best way to handle this execution, because the tool has a caching mechanism and is smarter than everyone but itself to determine which dependencies to cache.

I have to dig deeper into the engine to know what are the things needed for `buildkit` to skip executing a step, but IMO I would advise against it and provision the necessary context for each tool to decide.

## Making builds faster

As you said, sometimes this does not depend on Dagger, and it may depend on the caching/bundling strategy of the tool itself. There are for example in React, a LOT of caching and building strategies being implemented by different bundlers, but this is mostly because React is not a framework, but a library, so it allows using external tools to bundle, build, cache, etc.

From what I can gather, Angular is by itself a big entity that takes care of everything, even building. Here's [a comment](https://www.reddit.com/r/Angular2/comments/10by6fg/comment/j4ea0o9/?utm_source=share&utm_medium=web2x&context=3) from an Angular core member on why Angular cannot benefit from using some external tools for building.

In the case of Angular, you are already mounting the Angular cache, so I'm not sure if we can improve building times with Dagger itself. What I suggest is to try to build Angular without Dagger and with Dagger. If it takes less time without Dagger, there's some config missing in the pipeline that doesn't let the Angular cache strategy to work correctly. If it takes the same amount, the optimization must be done from Angular itself, or using an external tool that Angular can benefit from.
I guess you already tried this though as you stated the `ng` build command is not smart. 

Please, LMK if I missed something!
